documentation/architecture.md

```diff
@@ -0,0 +1,247 @@
+(`stdout`, `stderr`) back to `cc-proxy` and receiving the input stream (`stdin`) for them.
+As all streams for all containers are going through one single serial channel, hyperstart
+prepend them with container specific sequence numbers. There are at most 3 sequence numbers
+per container process, one for each stream (`stdout`, `stderr`, `stdin`)
+
+#### Runtime
+
+`cc-oci-runtime` is an OCI compatible container runtime and is responsible for handling all
+commands specified by [the OCI runtime specification](https://github.com/opencontainers/runtime-spec)
+and launching `cc-shim` instances.
+
+Here we will describe how `cc-oci-runtime` handles the most important OCI commands.
+
+##### `create`
+
+When handling the OCI `create` command, `cc-oci-runtime` goes through the following steps:
+
+1. Create the container namespaces (Only the network and mount namespaces are currently supported),
+according to the container OCI configuration file.
+2. Spawn the `cc-shim` process and have it wait on a couple of temporary pipes for:
+   * A `cc-proxy` created file descriptor (one end of a socketpair) for the shim to connect to.
+   * The container `hyperstart` sequence numbers for at most 3 I/O streams (`stdout`, `stderr`, `stdin`). `hyperstart` uses
+   those sequence numbers to multiplex all streams for all processes through one serial interface (The virtio I/O serial
+   one).
+3. Run all the [OCI hooks](https://github.com/opencontainers/runtime-spec/blob/master/config.md#hooks) in the container namespaces,
+as described by the OCI container configuration file.
+4. [Set the container networking up](https://github.com/01org/cc-oci-runtime/blob/topic/documentation/documentation/architecture.md#networking).
+This must happen after all hooks are done as one of them is potentially setting
+the container networking namespace up.
+5. Create the virtual machine running the container process. The VM `systemd` instance will spawn the `hyperstart` daemon.
+6. Wait for `hyperstart` to signal that it is ready.
+7. Send the pod creation command to `hyperstart`. The `hyperstart` pod is the container process sandbox.
+8. Send the allocateIO command to the proxy, for getting the `hyperstart` I/O sequence numbers described in step 2.
+9. Pass the `cc-proxy` socketpair file descriptor, and the I/O sequence numbers to the listening cc-shim process through the dedicated pipes.
+10. The `cc-shim` instance is put to a stopped state to prevent it from doing I/O before the container is started.
+
+![Docker create](https://github.com/01org/cc-oci-runtime/blob/topic/documentation/documentation/create.png)
+
+At that point, the container sandbox is created in the virtual machine and `cc-shim` is stopped on the host.
+But the container process itself is not yet running as one need to call `docker start` for actually starting it.
+
+##### `start`
+
+On namespace containers `start` mostly launches the container process on its namespaces.
```

Comment:
"`start` launches a traditional Linux container process in its own set of namespaces."